dali/kernels/reduce/reduce_drop_dims.h

```diff
@@ -0,0 +1,245 @@
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+#include <stdexcept>
+#include "dali/core/util.h"
+#include "dali/core/host_dev.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+/**
+ * Calculates a flat index in a tensor with reduced dimensions based on the
+ * index in original tensor.
+ *
+ * This is used in calculation of standard deviation when subtracting mean.
+ *
+ * @details When reducing non-adjacent dimensions, the outer reduction will see
+ * all inner dimensions flattened - reduced and non-reduced alike. This flat index
+ * cannot be used to index the the tensor of means, which has more reduced dimensions.
+ * Example - calculating mean/stddev of all pixel brightness values in a mutli-channel video:
+ * input layout FHWC
+ * output layout 1HW1
+ * When calculating variance in F, we still see HWC as the (fused) inner dimension, but it is
+ * different than HW1 in the tensor of means. This class implements this kind of reindexing.
+ *
+ * The reindexing is done by either dividing and multiplying by old/new strides or by takind modulo.
+ */
+struct DropDims {
+  static constexpr int kMaxDims = 4;
+
+  int64_t div[kMaxDims];
+  int64_t mul[kMaxDims];
+  int64_t mod[kMaxDims];
+  int start = 2 * kMaxDims;
+
+  DropDims() = default;
+
+  /**
+   * Collapses adjacent groups of reduced/non-reduced dimensions.
+   * The input can have up to 64 dimensions, however, the result must havfe
+   * at most 8 groups of reduced/non-reduced dimensions.
+   *
+   * Example:
+   * ```
+   * in_shape:  [ 2, 3, 4, 5, 6 ]
+   * reduced:     ^        ^  ^
+   * axis_mask = 0b11001  (looks reversed - LSB is dim 0)
+   *
+   * out_shape: [ 2, 12, 30 ]
+   * reduced:     ^       ^
+   * out_mask = 0b101
+   * ```
+   */
+  template <typename Indices>
+  static int simplify(int64_t *out_shape, unsigned &out_mask,
+                      const Indices &in_shape, uint64_t axis_mask) {
+    int dims = size(in_shape);
+    int d = 0;
+    out_shape[0] = in_shape[0];
+    bool prev = axis_mask & 1;
+    out_mask = prev ? 1u : 0u;
+    for (int i = 1; i < dims; i++) {
+      if (in_shape[i] == 1)
+        continue;
+      bool flag = (axis_mask >> i) & 1;
+      if (flag != prev) {
+        d++;
+        if (d > 2*kMaxDims)
+          throw std::out_of_range("Maximum number of dimension groups exceeded");
+        out_shape[d] = in_shape[i];
+        out_mask |= (flag ? 1u : 0u) << d;
+      } else {
+        out_shape[d] *= in_shape[i];
+      }
+      prev = flag;
+    }
+    d++;
+    if (d > 2*kMaxDims)
+      throw std::out_of_range("Maximum number of dimension groups exceeded");
+
+    return d;
+  }
+
+  /**
+   * @brief Initializes reindexing given shape and mask.
+   */
+  template <typename Indices>
+  DropDims(const Indices &in_shape, uint64_t reduced_axes) {
+    int64_t shape[kMaxDims];
+    unsigned axis_mask;
+    int d = simplify(shape, axis_mask, in_shape, reduced_axes);
+    start = 2 * kMaxDims;
```

Comment:
Can we get `axis_mask` like `b10001`?

Reply:
No. It will always be alternating bit pattern.